rlp/internal/rlpstruct/rlpstruct.go

```diff
@@ -0,0 +1,213 @@
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+// Package rlpstruct implements struct processing for RLP encoding/decoding.
+//
+// In particular, this package handles all rules around field filtering,
+// struct tags and nil value determination.
+package rlpstruct
+
+import (
+	"fmt"
+	"reflect"
+	"strings"
+)
+
+// Field represents a struct field.
+type Field struct {
+	Name     string
+	Index    int
+	Exported bool
+	Type     Type
+	Tag      string
+}
+
+// Type represents the attributes of a Go type.
+type Type struct {
+	Name      string
+	Kind      reflect.Kind
+	IsEncoder bool  // whether type implements rlp.Encoder
+	IsDecoder bool  // whether type implements rlp.Decoder
+	Elem      *Type // non-nil for Kind values of Ptr, Slice, Array
+}
+
+// defaultNilValue determines whether a nil pointer to t encodes/decodes
+// as an empty string or empty list.
+func (t Type) DefaultNilValue() NilKind {
+	k := t.Kind
+	if isUint(k) || k == reflect.String || k == reflect.Bool || isByteArray(t) {
+		return NilKindString
+	}
+	return NilKindList
+}
+
+// NilKind is the RLP value encoded in place of nil pointers.
+type NilKind uint8
+
+const (
+	NilKindString NilKind = 0x80
+	NilKindList   NilKind = 0xC0
+)
+
+// Tags represents struct tags.
+type Tags struct {
+	// rlp:"nil" controls whether empty input results in a nil pointer.
+	// nilKind is the kind of empty value allowed for the field.
+	NilKind NilKind
+	NilOK   bool
+
+	// rlp:"optional" allows for a field to be missing in the input list.
+	// If this is set, all subsequent fields must also be optional.
+	Optional bool
+
+	// rlp:"tail" controls whether this field swallows additional list elements. It can
+	// only be set for the last field, which must be of slice type.
+	Tail bool
+
+	// rlp:"-" ignores fields.
+	Ignored bool
+}
+
+// TagError is raised for invalid struct tags.
+type TagError struct {
+	StructType string
+
+	// These are set by this package.
+	Field string
+	Tag   string
+	Err   string
+}
+
+func (e TagError) Error() string {
+	field := "field " + e.Field
+	if e.StructType != "" {
+		field = e.StructType + "." + e.Field
+	}
+	return fmt.Sprintf("rlp: invalid struct tag %q for %s (%s)", e.Tag, field, e.Err)
+}
+
+// ProcessFields filters the given struct fields, returning only fields
+// that should be considered for encoding/decoding.
+func ProcessFields(allFields []Field) ([]Field, []Tags, error) {
+	lastPublic := lastPublicField(allFields)
+
+	// Gather all exported fields and their tags.
+	var fields []Field
+	var tags []Tags
```

Comment:
```suggestion
	fields := make([]Field, 0)
	tags := make([]Tags, 0)
```
suggestion to fix prealloc problem from test-linter

Reply:
Thanks. @hqjang-pepper.